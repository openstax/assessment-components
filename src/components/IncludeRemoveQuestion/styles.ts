import { colors } from '../../theme';
import styled from "styled-components";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';


export const StyledContainer = styled.div`
  display: flex;
  width: fit-content;
`;

export const StyledButton = styled.button`
    width: 74px;
    height: 87px;
    color: white;
    border: none;

    &:hover {
      box-shadow: 0px 1px 4px 0px #00000066;
    }
    
    span {
        font-family: "HelveticaNeue";
        line-height: 13px;
        font-size: 14px;
        font-weight: bold;
    }

    &.include {
        background-color: ${colors.palette.mediumBlue};
        padding-top: 14px;
    }

    &.remove {
        background-color: ${colors.palette.orange};
        padding-top: 14px;
    }

    &.details {
        background-color: ${colors.palette.neutralDarker};
    }
`;

export const StyledIcon = styled(FontAwesomeIcon)`
    border-radius: 50%;
    background-color: ${colors.palette.white};
    vertical-align: middle;
    position: relative;
    bottom: 5px;
    font-weight: 900;
    line-height: 13.8px;


    &.include {
        color: ${colors.palette.mediumBlue};
    }

    &.remove {
        color: ${colors.palette.orange};
    }

    &.details {
        color: ${colors.palette.neutralDarker};
    }
`;
