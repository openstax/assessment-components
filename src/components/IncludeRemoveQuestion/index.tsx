import React from "react";
import { faPlus, faMinus, faEllipsisH } from "@fortawesome/free-solid-svg-icons";
import { StyledContainer, StyledButton, StyledIcon } from "./styles";

export interface IncludeRemoveQuestionProps {
  // Prop that defines the variant of the button
  buttonVariant: 'include' | 'remove';
  // Method invoked when the include button is clicked
  onIncludeHandler: () => void;
    // Method invoked when the remove button is clicked
  onRemoveHandler: () => void;
}

export const IncludeRemoveQuestion = ({buttonVariant, onIncludeHandler, onRemoveHandler }: IncludeRemoveQuestionProps) => {

    const buttonIcon = React.useMemo(() => buttonVariant === 'include' ? faPlus : faMinus, [buttonVariant]);
    const onClickHandler = (variant: 'include' | 'remove') => { 
        switch (variant) {
            case 'include':
              onIncludeHandler();
              break;
            case 'remove':
              onRemoveHandler();
              break;
            default:
              break;
          }
    };

    const generateButtonText = (string: string) => {
      return string.charAt(0).toUpperCase() + string.slice(1) + ' question';
    };
    
    return (
        <StyledContainer>
            <StyledButton className={buttonVariant} onClick={() => onClickHandler(buttonVariant)} aria-label={buttonVariant}>
                <StyledIcon className={buttonVariant} icon={buttonIcon} aria-label={buttonVariant + ' question'} border size="lg" />
                <span>{generateButtonText(buttonVariant)}</span>
            </StyledButton>
            <StyledButton className="details" aria-label="details">
                <StyledIcon className="details" icon={faEllipsisH} border size="lg"/>
                <span>Details</span>
            </StyledButton>
        </StyledContainer>
    );
}; 