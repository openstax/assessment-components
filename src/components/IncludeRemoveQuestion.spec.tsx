import renderer from 'react-test-renderer';
import { IncludeRemoveQuestion } from './IncludeRemoveQuestion/index';

describe('IncludeRemoveQuestion', () => {
  it.each`
    buttonVariant
    ${'include'}
    ${'remove'}
  `('matches snapshot', ({ buttonVariant }: { buttonVariant: 'include' | 'remove' }) => {
    const mockIncludeHandler = jest.fn();
    const mockRemoveHandler = jest.fn();
    const mockOnClickDetails = jest.fn();
    const component = renderer.create(
      <IncludeRemoveQuestion 
        buttonVariant={buttonVariant} 
        onIncludeHandler={mockIncludeHandler} 
        onRemoveHandler={mockRemoveHandler}
        onClickDetails={mockOnClickDetails}
      />
    );
  
    renderer.act(() => {
        component.root.findAllByType('button')[0].props.onClick();
        component.root.findAllByType('button')[1].props.onClick();
    });
    
    expect(buttonVariant === 'include' ? mockIncludeHandler : mockRemoveHandler).toHaveBeenCalled();
    expect(mockOnClickDetails).toHaveBeenCalled();
    expect(component).toMatchSnapshot();
  });
});
