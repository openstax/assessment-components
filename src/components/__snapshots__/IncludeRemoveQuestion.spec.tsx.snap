// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncludeRemoveQuestion matches snapshot 1`] = `
<div
  className="sc-bczRLJ fXHYVE"
>
  <button
    aria-label="details"
    className="sc-gsnTZi gmAXwu include"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      aria-label="include question"
      className="svg-inline--fa fa-plus fa-border fa-lg sc-dkzDqf bHPCba include"
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Include question
    </span>
  </button>
  <button
    aria-label="details button"
    className="sc-gsnTZi gmAXwu details"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-ellipsis fa-border fa-lg sc-dkzDqf bHPCba details"
      data-icon="ellipsis"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M120 256c0 30.9-25.1 56-56 56s-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56zm160 0c0 30.9-25.1 56-56 56s-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56zm104 56c-30.9 0-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56s-25.1 56-56 56z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Details
    </span>
  </button>
</div>
`;

exports[`IncludeRemoveQuestion matches snapshot 2`] = `
<div
  className="sc-bczRLJ fXHYVE"
>
  <button
    aria-label="details"
    className="sc-gsnTZi gmAXwu remove"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      aria-label="remove question"
      className="svg-inline--fa fa-minus fa-border fa-lg sc-dkzDqf bHPCba remove"
      data-icon="minus"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Remove question
    </span>
  </button>
  <button
    aria-label="details button"
    className="sc-gsnTZi gmAXwu details"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-ellipsis fa-border fa-lg sc-dkzDqf bHPCba details"
      data-icon="ellipsis"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M120 256c0 30.9-25.1 56-56 56s-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56zm160 0c0 30.9-25.1 56-56 56s-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56zm104 56c-30.9 0-56-25.1-56-56s25.1-56 56-56s56 25.1 56 56s-25.1 56-56 56z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Details
    </span>
  </button>
</div>
`;
